### DISCUSSION TITLE
Using CCF as computational Platform.

### DISCUSSION THREAD

**Discussion by aamar-shahzad** [microsoft/CCF#5881] [Q&A]
Hi Folks,
here is Aamar, I am using the CCF as DLT in AI for computation due to its reputation with Hardware security feature.
But I am facing a challenge in using CCF for multiparty computation in AI for server level computation as I am using it as server less technology but due its computational limit of JS execution timeout failure. I am not able to run a computational extensive function in any case like asyn execuation.
Is there way to change the code from HTTP to gRPC protocol to make it more usable in real time use cases.
Looking for you comments on 

{},s 

aamar

**Comment by eddyashton** [comment]
Hi Aamar.

I think there are several questions here:
- Hitting JS execution timeouts in long-lived transactions. These limits can be extended as documented [here](https://microsoft.github.io/CCF/main/governance/js_runtime.html) to support longer operations, though you may also need to extend other timeouts (consensus elections, node forwarding caps) to prevent long-running operations stalling the network entirely.
- Running async functions. This is not possible - CCF operations are currently blocking, synchronous RPC handlers. To provide a meaningful guarantee, the operation's result should be the deterministic result of the given request and the current KV state, so we do not believe there is any reason to do common async work (ie - if you wanted to call out to an external service during a handler, the operation would no longer be a transaction). There's some discussion [here](https://github.com/microsoft/CCF/discussions/4331) around supporting async language features, but this would still be executed synchronously/inline - it would allow use of JS libraries which used async, but not arbitrary IO.
- Using gRPC rather than HTTP. There is [some support](https://github.com/microsoft/CCF/tree/main/src/endpoints/grpc) for gRPC already, although it is not currently part of the public API. The LSKV project built on CCF uses this to provide a subset of the etcd API. We'd like to improve and ship this as a feature, but there are some open questions around how streams should map to transactions, and how this fits into the current handler API. There's no immediate plan to complete this work, but if you can expand on your gRPC requirements we can consider them.

Can you expand on the kind of computations you'd like to do within a CCF node? AI tasks generally require specialised hardware (GPUs), so may not be suited for the SGX/SEV-SNP hardware that CCF (currently) supports.

**Reply by aamar-shahzad** [reply]
Thank you, Eddy, for your response.
 Your explanation of the insights and capabilities of CCF has been very helpful in guiding my next steps. I am expanding the CCF framework to include DLT services for AI model aggregation in federated learning, particularly focusing on model storage. This is being implemented as serverless architecture to mitigate issues related to central server failure. Here's what I've achieved so far:

- I've successfully stored some CNN models, but I haven't tried with LLM models yet.
- I'm able to retrieve AI models upon request from clients.
- However, I'm facing challenges with model aggregation, which requires operations to be extended in duration. I attempted to resolve this by using asynchronous functions, but I'm not entirely confident in their correct execution. Additionally, I've been noticing JS timeouts in the logs.
- I've tried to increase the JavaScript computation timeout by changing the `open_session_hard TIMEOUT`, but this hasn't been successful.
- If resolved, this could be a viable solution for replacing other blockchain technologies for AI.
- I am considering using gRPC as a server instead of RESTful services to enhance reliability and efficiency for AI tasks, as seen in other open-source frameworks.

Here is what I want:
- adding the asynchronous execution for longer task.
- Option of switching  between gRPC and Restful.
- Increasing JS timeout limit with configuration.


`Can you expand on the kind of computations you'd like to do within a CCF node? AI tasks generally require specialized hardware (GPUs), so may not be suited for the SGX/SEV-SNP hardware that CCF (currently) supports.`
Idea is to use CCF as services for storing of model and aggregation  not as model training platform and for that it not required any GPU or special CPU.

**Reply by achamayou** [reply]
@aamar-shahzad to extend the JS execution timeout, you will need to [send and pass a governance proposal](https://microsoft.github.io/CCF/main/governance/proposals.html) to update `max_execution_time_ms` (as described [here](https://microsoft.github.io/CCF/main/governance/js_runtime.html)).
